src/cider/type/plan/CaseExpr.cpp

```diff
@@ -0,0 +1,277 @@
+  }
+}
+
+void CaseExpr::collect_rte_idx(std::set<int>& rte_idx_set) const {
+  for (auto p : expr_pair_list) {
+    p.first->collect_rte_idx(rte_idx_set);
+    p.second->collect_rte_idx(rte_idx_set);
+  }
+  if (else_expr != nullptr) {
+    else_expr->collect_rte_idx(rte_idx_set);
+  }
+}
+
+void CaseExpr::collect_column_var(
+    std::set<const ColumnVar*, bool (*)(const ColumnVar*, const ColumnVar*)>& colvar_set,
+    bool include_agg) const {
+  for (auto p : expr_pair_list) {
+    p.first->collect_column_var(colvar_set, include_agg);
+    p.second->collect_column_var(colvar_set, include_agg);
+  }
+  if (else_expr != nullptr) {
+    else_expr->collect_column_var(colvar_set, include_agg);
+  }
+}
+
+void CaseExpr::check_group_by(
+    const std::list<std::shared_ptr<Analyzer::Expr>>& groupby) const {
+  for (auto p : expr_pair_list) {
+    p.first->check_group_by(groupby);
+    p.second->check_group_by(groupby);
+  }
+  if (else_expr != nullptr) {
+    else_expr->check_group_by(groupby);
+  }
+}
+
+JITExprValue& CaseExpr::codegen(CodegenContext& context) {
+  JITFunction& func = *context.getJITFunction();
+  const auto& expr_pair_list = get_expr_pair_list();
+  const auto& else_expr = get_else_ref();
+  CHECK_GT(expr_pair_list.size(), 0);
+  const auto case_ti = get_type_info();
+  if (case_ti.is_integer() || case_ti.is_time() || case_ti.is_decimal() ||
+      case_ti.is_fp() || case_ti.is_boolean()) {
+    const auto type =
+        case_ti.is_decimal() ? decimal_to_int_type(case_ti) : case_ti.get_type();
+    JITValuePointer value = func.createVariable(getJITTag(type), "case_when_value_init");
+    *value = func.createLiteral(getJITTag(type), 0);
+    JITValuePointer null = func.createVariable(JITTypeTag::BOOL, "case_when_null_init");
+    *null = func.createLiteral(JITTypeTag::BOOL, false);
+    JITValuePointer is_case = func.createVariable(JITTypeTag::BOOL, "is_case_init");
+    *is_case = func.createLiteral(JITTypeTag::BOOL, false);
+    for (const auto& expr_pair : expr_pair_list) {
+      func.createIfBuilder()
+          ->condition([&]() {
+            cider::exec::nextgen::utils::FixSizeJITExprValue cond(
+                expr_pair.first->codegen(context));
+            auto condition = !cond.getNull() && cond.getValue();
+            return condition;
+          })
+          ->ifTrue([&]() {
+            cider::exec::nextgen::utils::FixSizeJITExprValue then_jit_expr_value(
+                expr_pair.second->codegen(context));
+            *value = *then_jit_expr_value.getValue();
+            *null = *then_jit_expr_value.getNull();
+            *is_case = func.createLiteral(JITTypeTag::BOOL, true);
+          })
+          ->build();
+    }
+    func.createIfBuilder()
+        ->condition([&]() { return !is_case; })
+        ->ifTrue([&]() {
+          cider::exec::nextgen::utils::FixSizeJITExprValue else_jit_expr_value(
+              else_expr->codegen(context));
+          *value = *else_jit_expr_value.getValue();
```

Comment:
`value = else_jit_expr_value.getValue();` would be fine? `*` here will bring a memory copy I guess?

Reply:
Yes, this will do a memory copy. Use JITValuePointer.replace instant of * copy.